test/observer/permissions.dart

```diff
@@ -0,0 +1,79 @@
+  final List<Permission> _checkedPermissions = [];  // The permissions which were checked.
+  List<Permission> get checkedPermissions => UnmodifiableListView(_checkedPermissions);
+  int _openSettingsRequests = 0;  /// The amount of attempts to open the permission settings.
+  int get openSettingsRequests => _openSettingsRequests;
+
+  /// Whether an attempt to open the settings should succeed.
+  bool isOpeningSettingsSuccessful = true;
+  
+  /// Permissions whose values are specified, all other permissions are implicitly granted.
+  final Map<Permission, Future<PermissionStatus> Function()> _specifiedPermissions = {};
+
+  /// Create a new permissions observer, which automatically
+  /// unregisters any previously created observer.
+  PermissionsChannelObserver(TestWidgetsFlutterBinding binding) {
+    binding.defaultBinaryMessenger.setMockMethodCallHandler(_channel, (call) async {
+      switch (call.method) {
+        case 'requestPermissions':
+          final Map<int, int> permissionStatuses = {};
+          for (final int permissionValue in List<int>.from(call.arguments)) {
+            final permission = Permission.byValue(permissionValue);
+            _requestedPermissions.add(permission);
+            final permissionStatus = await getStatus(permission);
+            permissionStatuses[permissionValue] = permissionStatus.index;
+          }
+          return permissionStatuses;
+        case 'checkPermissionStatus':
+          final Permission permission = Permission.byValue(call.arguments as int);
+          _checkedPermissions.add(permission);
+          final status = await getStatus(permission);
+          return status.index;
+        case 'openAppSettings':
+          _openSettingsRequests++;
+          return isOpeningSettingsSuccessful;
+      }
+      return null;  // Ignore unimplemented method calls
+    });
+  }
+  
+  /// Get the status of the given [permission].
+  Future<PermissionStatus> getStatus(Permission permission) {
+    return (_specifiedPermissions[permission] ?? () async => PermissionStatus.granted)();
+  }
+
+  /// Set the status value of the [permission] to the result of the
+  /// [resolvable], which will be resolved when the permission status is
+  /// first requested.
+  void setPermissionResolvable(Permission permission, Future<PermissionStatus> Function() resolvable) {
+    _specifiedPermissions[permission] = resolvable;
+  }
+  
+  /// Set the [status] for the [permission].
+  void setPermission(Permission permission, PermissionStatus status) {
+    if (status == PermissionStatus.granted) {
+      _specifiedPermissions.remove(permission);
+    } else {
+      setPermissionResolvable(permission, () async => status);
+    }
+  }
+}
+
+extension PermissionObserverExtension on WidgetTester {
+  /// Create a new PermissionsChannelObserver, which overwrites the default one
+  /// for this test.
+  PermissionsChannelObserver overwritePermissionObserver() => PermissionsChannelObserver(binding);
```

Comment:
Why do we even need this? Why not just `permissionsObserver = PermissionsChannelObserver()`;

Reply:
Unfortunately, extensions can't declare fields.

Reply:
Ah, if you mean why we need the extension at all, it's to avoid having to write:
```dart
final permissionsObserver = PermissionsChannelObserver(tester.binding);
```

So it's not necessary, but it makes the code look better and also self documents whats happening (overwriting the default permission observer).

Reply:
That's an inherentent action of creating an observer, so I'm for removing this method, since it's an equivalent of a constructor call